ophyd/status.py

```diff
@@ -1134,3 +1143,299 @@ def wait(status, timeout=None, *, poll_rate="DEPRECATED"):
+        event_type=None,
+        timeout: float = None,
+        settle_time: float = 0,
+    ):
+        self._signal = signal
+        if not isinstance(transitions, list):
+            raise ValueError(
+                f"Transitions must be a list of values. Received: {transitions}"
+            )
+        self._transitions = transitions
+        self._index = 0
+        self._strict = strict
+        self._raise_states = raise_states if raise_states else []
+        super().__init__(
+            device=signal,
+            callback=self._compare_callback,
+            timeout=timeout,
+            settle_time=settle_time,
+            event_type=event_type,
+            run=run,
+        )
+
+    def _compare_callback(self, old_value, value, **kwargs) -> bool:
+        """Callback for subscription Status"""
+        if value in self._raise_states:
+            self.set_exception(
+                ValueError(
+                    f"Transition Status for {self._signal.name} resulted in a value: {value}. "
+                    f"marked to raise {self._raise_states}. Expected transitions: {self._transitions}."
+                )
+            )
+            return False
+        if self._index == 0:
+            if value == self._transitions[0]:
+                self._index += 1
+        else:
+            if self._strict:
+                if (
+                    old_value == self._transitions[self._index - 1]
+                    and value == self._transitions[self._index]
+                ):
+                    self._index += 1
+            else:
+                if value == self._transitions[self._index]:
+                    self._index += 1
+        return self._is_finished()
+
+    def _is_finished(self) -> bool:
+        """Check if the status is finished"""
+        return self._index >= len(self._transitions)
+
+
+class AndAllStatus(DeviceStatus):
+    """
+    A status that combines mutiple status objects in a list using logical and.
+    The status is finished when all status objects in the list are finished.
+    If any status object fails, the combined status will also fail and
+    set the exception from the first failed status on all sub-statuses.
+
+    Parameters
+    ----------
+    device: Device
+    status_list: A list of StatusBase or DeviceStatus objects to combine.
+    """
+
+    def __init__(
+        self, device: Device, status_list: list[StatusBase | DeviceStatus], **kwargs
+    ):
+        self.status_list = status_list
+        super().__init__(device=device, **kwargs)
+        self._trace_attributes["all"] = [
+            st._trace_attributes for st in self.status_list
+        ]
+
+        def inner(status):
+            with self._lock:
+                if self._externally_initiated_completion:
+                    return
+
+                # Return if status is already done..
+                if self.done:
+                    return
+
+                for st in self.status_list:
```

Comment:
Why loop through the list?  Isn't the `status` object passed in the most recently changed one and we can just react to that?